app/src/main/java/amritansh/tripathi/com/searchflix/presentation/movieList/paging_lib/MoviesDataSource.kt

```diff
@@ -0,0 +1,68 @@
+package amritansh.tripathi.com.searchflix.presentation.movieList.paging_lib
+
+import amritansh.tripathi.com.searchflix.domain.usecases.GetPopularMoviesUseCase
+import amritansh.tripathi.com.searchflix.network.Movie
+import amritansh.tripathi.com.searchflix.presentation.models.UiState
+import android.arch.lifecycle.MutableLiveData
+import android.arch.paging.PageKeyedDataSource
+import io.reactivex.Completable
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.functions.Action
+import io.reactivex.schedulers.Schedulers
+
+class MoviesDataSource(
+        private val getPopularMoviesUseCase: GetPopularMoviesUseCase,
+        private val compositeDisposable: CompositeDisposable) : PageKeyedDataSource<Int, Movie>() {
+
+    var state = MutableLiveData<UiState>()
+    var retryComplete: Completable? = null
+
+    override fun loadInitial(params: LoadInitialParams<Int>, callback: LoadInitialCallback<Int, Movie>) {
+        state.postValue(UiState.LOADING)
+        val pageInit = 1
+        compositeDisposable.add(
+                getPopularMoviesUseCase.getPopularMovieUseCase(pageInit, params.requestedLoadSize)
+                        .subscribe({
+                            state.postValue(UiState.COMPLETE)
+                            callback.onResult(it.result, null, 2)
+                        }, {
+                            state.postValue(UiState.ERROR)
+                            setRetry(Action { loadInitial(params, callback) })
+                        })
```

Comment:
You can remove this block from the `.add` and call `.addTo(compositeDisposable)` I guess